### PR TITLE
[expo-updates] Updates state machine for iOS and Android, E2E tests

### PR DESCRIPTION
**Pull request by douglowder**
# Why

Implement state machine on Android and iOS.

# How

- New state machine classes on Android and iOS
- Integration with UpdatesController/AppController
- State machine change events sent to JS
- Test hook in `@expo/use-updates` to allow E2E tests
- E2E tests for both platforms

# Test Plan

- Manual testing on both platforms
- Unit tests for state machine on both platforms
- New E2E tests for state machine feature
- Existing E2E tests and existing `expo-updates` functionality should be unaffected

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-updates/android/build.gradle [hunk 1]
```diff
@@ -6,7 +6,7 @@ apply plugin: 'maven-publish'
 group = 'host.exp.exponent'
 version = '0.17.0'
 
-def ex_updates_native_debug = System.getenv("EX_UPDATES_NATIVE_DEBUG") == "1" ? "true" : "false"
+def ex_updates_native_debug = System.getenv("EX_UPDATES_NATIVE_DEBUG") == "1" ? "true" : "true"
```
Comment (douglowder) [inline packages/expo-updates/android/build.gradle:9]:
```suggestion
def ex_updates_native_debug = System.getenv("EX_UPDATES_NATIVE_DEBUG") == "1" ? "true" : "false"
```

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.kt [hunk 2]
```diff
@@ -66,7 +68,7 @@ class UpdatesModule(
         constants["runtimeVersion"] = updatesServiceLocal.configuration.runtimeVersion ?: ""
         constants["checkAutomatically"] = updatesServiceLocal.configuration.checkOnLaunch.toJSString()
         constants["channel"] = updatesServiceLocal.configuration.requestHeaders["expo-channel-name"] ?: ""
-        constants["nativeDebug"] = BuildConfig.EX_UPDATES_NATIVE_DEBUG
+        constants["nativeDebug"] = true
```
Comment (douglowder) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.kt:71]:
```suggestion
        constants["nativeDebug"] = BuildConfig.EX_UPDATES_NATIVE_DEBUG
```

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesPackage.kt [hunk 3]
```diff
@@ -29,7 +29,7 @@ import expo.modules.updates.UpdatesController
  * applicable environments.
  */
 class UpdatesPackage : Package {
-  private val useNativeDebug = BuildConfig.EX_UPDATES_NATIVE_DEBUG
+  private val useNativeDebug = BuildConfig.EX_UPDATES_NATIVE_DEBUG || true
```
Comment (douglowder) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesPackage.kt:32]:
```suggestion
  private val useNativeDebug = BuildConfig.EX_UPDATES_NATIVE_DEBUG
```

File: packages/expo-structured-headers/ios/Tests/EXStructuredHeadersParserTests.m [hunk 4]
```diff
@@ -106,6 +106,7 @@ - (EXStructuredHeadersParserFieldType)fieldTypeWithString:(NSString *)string
     return EXStructuredHeadersParserFieldTypeItem;
   } else {
     XCTAssert(NO, @"unexpected header_type");
+    return EXStructuredHeadersParserFieldTypeDictionary;
```
Comment (douglowder) [inline packages/expo-structured-headers/ios/Tests/EXStructuredHeadersParserTests.m:109]: This fixes a compile issue with Xcode 14 when building `native-tests`.

File: packages/expo-updates/ios/EXUpdates/UpdatesUtils.swift [hunk 5]
```diff
@@ -111,6 +239,10 @@ public final class UpdatesUtils: NSObject {
     }
   }
 
+  internal static func sendStateEvent(_ type: UpdatesStateEventType, body: [String: Any] = [:]) {
+    AppController.sharedInstance.stateMachine.processEvent(UpdatesStateEvent(type: type, body: body))
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/UpdatesUtils.swift:243]: Same comment as android: we should separate out the concerns of these classes. Module/utils should call into controller, controller should hold business logic and update its state machine, and return back to the module/utils.

File: packages/expo-updates/ios/EXUpdates/AppLoader/AppLoaderTask.swift [hunk 6]
```diff
@@ -19,7 +18,10 @@ public protocol AppLoaderTaskDelegate: AnyObject {
    * AppLoaderTask proceed as usual.
    */
   func appLoaderTask(_: AppLoaderTask, didLoadCachedUpdate update: Update) -> Bool
-  func appLoaderTask(_: AppLoaderTask, didStartLoadingUpdate update: Update)
+  func appLoaderTask(_: AppLoaderTask, didStartCheckingForUpdate body: [String: Any])
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/AppLoader/AppLoaderTask.swift:21]:
```suggestion
  func appLoaderTask(_: AppLoaderTask, didStartCheckingForRemoteUpdate body: [String: Any])
```

(and similar naming changes for new delegate methods below as was suggested on android)

File: packages/expo-updates/ios/EXUpdates/AppLoader/AppLoaderTask.swift [hunk 7]
```diff
@@ -497,3 +540,7 @@ public final class AppLoaderTask: NSObject {
     }
   }
 }
+
+// swiftlint:enable closure_body_length
+// swiftlint:enable force_unwrapping
+// swiftlint:enable superfluous_else
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/AppLoader/AppLoaderTask.swift:546]: no need to re-renable swiftlint at the end of files. ditto for other files too

File: packages/expo-updates/ios/EXUpdates/UpdatesUtils.swift [hunk 8]
```diff
@@ -78,17 +52,171 @@ public final class UpdatesUtils: NSObject {
...
+            return
+          case is RollBackToEmbeddedUpdateDirective:
+            let body = [
+              "isRollBackToEmbedded": true
+            ]
+            block(body)
+            sendStateEvent(.checkCompleteAvailable, body: body)
+            return
+          default:
+            return handleCheckError(UpdatesUnsupportedDirectiveException(), block: block)
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/UpdatesUtils.swift:94]: we should throw this exception so that it uses the default error handling provided by `AsyncFunction` when an exception of this type is thrown (one that extends expo-modules-core Exception type). ditto with `handleFetchError`.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt [hunk 9]
```diff
@@ -299,7 +351,22 @@ class UpdatesController private constructor(
...
               params.putString("message", exception.message)
-              UpdatesUtils.sendEventToReactNative(reactNativeHost, UPDATE_ERROR_EVENT, params)
+              sendLegacyUpdateEventToJS(UPDATE_ERROR_EVENT, params)
+
+              val body: Map<String, Any> = mapOf(
+                "message" to (exception.message ?: "")
+              )
+              // Since errors can happen through a number of paths, we do these checks
+              // to make sure the state machine is valid
+              if (stateMachine.state == UpdatesStateValue.Idle) {
```
Comment (wschurman) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt:361]: nit: use `when` and make exhaustive for better compiler safety

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+
+  var json: [String: Any] {
+    return [
+      "isUpdateAvailable": self.isUpdateAvailable,
+      "isUpdatePending": self.isUpdatePending,
+      "isRollback": self.isRollback,
+      "isChecking": self.isChecking,
+      "isDownloading": self.isDownloading,
+      "isRestarting": self.isRestarting,
+      "latestManifest": self.latestManifest ?? NSNull(),
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:111]: I believe using NSNull in swift usually isn't necessary. Does using nil not work here? If it doesn't, add a comment indicating why we use NSNull

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+internal protocol UpdatesStateChangeEventSender {
+  func sendUpdateStateChangeEventToBridge(_ eventType: UpdatesStateEventType, body: [String: Any])
+}
+
+// MARK: - Enums
+
+/**
+ All the possible states the machine can take.
+ */
+internal enum UpdatesStateValue: String {
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:18]: private

**Review by wschurman — CHANGES_REQUESTED** [review]
This is a good addition, and one that I think we can use as the source of truth about the state of the controller/system as a whole in the near future (by asserting all state transitions are valid).

Left comments inline. I'm happy to expand upon anything that isn't clear over video chat/slack.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    }
+  }
+
+  /**
+   Make sure the state transition is allowed, and then update the state.
+   */
+  private fun transition(event: UpdatesStateEvent): Boolean {
+    val allowedEvents: List<UpdatesStateEventType> = updatesStateAllowedEvents[state] ?: listOf()
+    if (!allowedEvents.contains(event.type)) {
+      // Optionally put an assert here when testing to catch bad state transitions in E2E tests
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt:68]: For the JS API calls, we should not see any invalid state transitions, since they are queued in AsyncTask on Android and a dispatch queue on iOS. For the check and download that happens on launch, it's a bit more difficult, but we should definitely give some thought to how we can guarantee that only valid transitions will occur.

I did test this by adding an assert that would fire whenever an invalid transition is attempted, and crash the app. No such crashes are occurring in any of the E2E tests, so I have some confidence that they won't happen. However, we can add some chaos testing in E2E (press check button multiple times, etc.) to test this more thoroughly.

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift (hunk 11 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:18]: Nope this needs to be internal

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+
+  /**
+   In production, this is the AppController instance.
+   */
+  internal var changeEventSender: (any UpdatesStateChangeEventSender)?
+
+  /**
+   The current state
+   */
+  internal var state: UpdatesStateValue = .idle
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:163]: This must be internal so `AppController` can read it

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+      newContext.checkError = event.error
+    case .download:
+      newContext.isDownloading = true
+    case .downloadComplete:
+      newContext.isDownloading = false
+      newContext.downloadError = nil
+      newContext.latestManifest = event.manifest ?? context.latestManifest
+      newContext.downloadedManifest = event.manifest ?? context.downloadedManifest
+      newContext.isUpdatePending = newContext.downloadedManifest != nil
+      newContext.isUpdateAvailable = event.manifest != nil || newContext.isUpdateAvailable
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:248]: In the case of a rollback, you would get no manifest, but still have an update available (the rollback).

[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:248]: You could have the case where a manifest was downloaded and already ready to use to restart the app, and then the server removed the manifest. The already downloaded manifest is still valid.

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+  }
+
+  // MARK: - Static definitions of the state machine rules
+
+  /**
+   For a particular machine state, only certain events may be processed.
+   If the machine receives an unexpected event, an assertion failure will occur
+   and the app will crash.
+   */
+  static let updatesStateAllowedEvents: [UpdatesStateValue: [UpdatesStateEventType]] = [
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:285]: Not sure it makes a real perf difference with only 1 to 5 members, but yes we can do that :)

File: packages/@expo/use-updates/src/UseUpdates.ts [hunk 16]
```diff
@@ -206,3 +206,37 @@ export const useUpdates: () => UseUpdatesReturnType = () => {
...
+    latestManifest: null,
+    downloadedManifest: null,
+  });
+  useEffect(() => {
+    const subscription = addUpdatesStateChangeListener((event) => {
+      const state: { [key: string]: any } = {};
+      for (const key of event.fields) {
+        state[key] = event.values[key];
+      }
+      setLocalState((localState) => ({ ...localState, ...state }));
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/@expo/use-updates/src/UseUpdates.ts:237]: We are sending partial change events that only contain the changed properties, so a merge is required here.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesService.kt [hunk 17]
```diff
@@ -13,8 +13,8 @@ import java.io.File
 
 // these unused imports must stay because of versioning
 /* ktlint-disable no-unused-imports */
-import expo.modules.updates.UpdatesConfiguration
-import expo.modules.updates.UpdatesController
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesService.kt:17]: I think these were removed by spotlessApply, but I will check

File: packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt [hunk 18]
```diff
@@ -65,6 +65,11 @@ class LoaderTask(
     fun onCachedUpdateLoaded(update: UpdateEntity): Boolean
     fun onRemoteUpdateManifestResponseManifestLoaded(updateManifest: UpdateManifest)
     fun onSuccess(launcher: Launcher, isUpToDate: Boolean)
+
+    fun onCheckForUpdateStarted() {}
+    fun onCheckForUpdateFinished(body: Map<String, Any>) {}
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt:70]: I'm going with `onRemoteCheckForUpdateStarted/Finished` :)

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+  /**
+   * The context and logger are null in unit tests,
+   * but are instantiated in a production app.
+   */
+  private val logger = when (androidContext) {
+    null -> null
+    else -> UpdatesLogger(androidContext)
+  }
+
+  var changeEventSender: UpdatesStateChangeEventSender? = null
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt:19]: It can be null in unit tests.

File: packages/expo-updates/ios/EXUpdates/AppLoader/AppLoaderTask.swift (hunk 7 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/AppLoader/AppLoaderTask.swift:546]: actually SwiftLint now complains if you do not do this.

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift (hunk 10 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:93]: Since we are replacing the entire context on each state change, the NSNull is probably safe to remove, agreed.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt:60]: I added some tests in the state machine E2E where the check button is pressed multiple times in a row, then download pressed multiple times. Logs show that state transitions are occurring as expected. I'll add some more testing like this.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt [hunk 20]
```diff
@@ -65,6 +65,11 @@ class LoaderTask(
     fun onCachedUpdateLoaded(update: UpdateEntity): Boolean
     fun onRemoteUpdateManifestResponseManifestLoaded(updateManifest: UpdateManifest)
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt:71]: I think that should be a separate PR since it could potentially lead to regressions.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesInterface.kt [hunk 21]
```diff
@@ -23,6 +24,7 @@ interface UpdatesInterface {
   val directory: File?
   val databaseHolder: DatabaseHolder
   val fileDownloader: FileDownloader
+  val stateMachine: UpdatesStateMachine?
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesInterface.kt:27]: This is a great idea. It is also a large enough change that it deserves a separate task :)

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt:60]: I've added a new project in Linear for future work on the state machine and JS API, and added this and other tasks. https://linear.app/expo/project/expo-updates-js-api-2-de38dafeb7e4/ENG

File: packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt (hunk 20 above)
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt:71]: ENG-8997

File: packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt [hunk 22]
```diff
@@ -65,6 +65,11 @@ class LoaderTask(
...
     fun onSuccess(launcher: Launcher, isUpToDate: Boolean)
+
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/loader/LoaderTask.kt:73]: ENG-8997

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt:60]: ENG-8998

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesInterface.kt (hunk 21 above)
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesInterface.kt:27]: ENG-8995

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.kt [hunk 23]
```diff
@@ -256,9 +261,15 @@ object UpdatesUtils {
   }
 
   @Throws(ParseException::class)
-  fun parseDateString(dateString: String?): Date {
+  fun parseDateString(dateString: String?): Date? {
+    if (dateString == null) {
+      return null
+    }
```
Comment (wschurman) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.kt:267]: I think this is a somewhat risky change as it changes the defaulting of fields that are expected to be non-null. Instead, handle null at the callsites that pass nullable so that we don't need the `?: Date()` cases at all the callsites

```suggestion
  fun parseDateString(dateString: String): Date {
```

File: packages/expo-updates/android/src/main/java/expo/modules/updates/loader/RemoteUpdate.kt [hunk 24]
```diff
@@ -23,7 +23,7 @@ sealed class UpdateDirective(val signingInfo: SigningInfo?) {
       return when (val messageType = json.require<String>("type")) {
         "noUpdateAvailable" -> NoUpdateAvailableUpdateDirective(signingInfo)
         "rollBackToEmbedded" -> RollBackToEmbeddedUpdateDirective(
-          parseDateString(json.require<JSONObject>("parameters").require("commitTime")),
+          parseDateString(json.require<JSONObject>("parameters").require("commitTime")) ?: Date(),
```
Comment (wschurman) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/loader/RemoteUpdate.kt:26]: revert

File: packages/expo-updates/android/src/main/java/expo/modules/updates/manifest/LegacyUpdateManifest.kt [hunk 25]
```diff
@@ -107,7 +107,7 @@ class LegacyUpdateManifest private constructor(
         id = UUID.fromString(manifest.getReleaseId())
         val commitTimeString = manifest.getCommitTime()
         commitTime = try {
-          UpdatesUtils.parseDateString(commitTimeString)
+          UpdatesUtils.parseDateString(commitTimeString) ?: Date()
```
Comment (wschurman) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/manifest/LegacyUpdateManifest.kt:110]:
```
        commitTime = if (commitTimeString != null) {
          try {
            UpdatesUtils.parseDateString(commitTimeString) ?: Date()
          } catch (e: ParseException) {
            Log.e(TAG, "Could not parse commitTime", e)
            Date()
          }
        } else {
          Date()
        }
```

File: packages/expo-updates/android/src/main/java/expo/modules/updates/manifest/NewUpdateManifest.kt [hunk 26]
```diff
@@ -99,7 +99,7 @@ class NewUpdateManifest private constructor(
       val launchAsset = manifest.getLaunchAsset()
       val assets = manifest.getAssets()
       val commitTime: Date = try {
-        UpdatesUtils.parseDateString(manifest.getCreatedAt())
+        UpdatesUtils.parseDateString(manifest.getCreatedAt()) ?: Date()
```
Comment (wschurman) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/manifest/NewUpdateManifest.kt:102]: revert

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateContext.kt [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+      val map: MutableMap<String, Any> = mutableMapOf(
+        "isUpdateAvailable" to isUpdateAvailable,
+        "isUpdatePending" to isUpdatePending,
+        "isRollback" to isRollback,
+        "isChecking" to isChecking,
+        "isDownloading" to isDownloading,
+        "isRestarting" to isRestarting
+      )
+      if (latestManifest != null) {
+        map["latestManifest"] = latestManifest!!
```
Comment (wschurman) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateContext.kt:32]: By using `val` in the fields, no need for this non-null-assertion since the value is immutable. (ditto with others below)

```suggestion
        map["latestManifest"] = latestManifest
```

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt [hunk 28]
```diff
@@ -0,0 +1,199 @@
...
+  /**
+   * If a state change event is passed in, the JS sender
+   * is called with just the fields and values that changed.
+   * During a reset, this method is called with no event passed in,
+   * and then all the fields and the entire context are passed to the JS sender.
+   */
+  private fun sendChangeEventToJS(event: UpdatesStateEvent? = null) {
+    changeEventSender.sendUpdateStateChangeEventToBridge(
+      event?.type ?: UpdatesStateEventType.Restart,
+      context.json
```
Comment (wschurman) [inline packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt:76]: instead of passing JSON and then calling back into this class to convert to writable map, just pass the value of `paramsForJSEvent`. Even better, add a `toWritableMap` method to `UpdatesStateContext` so that the hard to follow `field.startsWith("is")` and `field.endsWith("Manifest")` are not necessary.

**Review by wschurman — CHANGES_REQUESTED** [review]

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt (hunk 28 above)
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt:76]: I will try that -- however I originally did things this way because I found that attempting to create WritableMaps didn't work unless the bridge was available, so things broke in unit tests. I'll have another look to see if it is possible to go directly to WritableMap.

File: packages/@expo/use-updates/src/UseUpdates.ts [hunk 29]
```diff
@@ -206,3 +206,33 @@ export const useUpdates: () => UseUpdatesReturnType = () => {
...
+
+/**
+ * Experimental hook to return the Updates state machine context maintained
+ * in native code.
+ *
+ * Eventually, this will be used to construct the information returned by `useUpdates()`.
+ *
+ * @returns A map of the state machine context.
+ */
+export const useUpdatesState: () => { [key: string]: any } = () => {
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/@expo/use-updates/src/UseUpdates.ts:218]: We will definitely do so once we build out the JS API

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.kt [hunk 30]
```diff
@@ -256,9 +261,15 @@ object UpdatesUtils {
...
     return try {
-      val formatter: DateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX", Locale.US)
+      val formatter: DateFormat = when (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+        true -> SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'X'", Locale.US)
+        else -> SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesUtils.kt:271]: Done. Leaving the ParseException catch in case a bad string is passed in :)

File: packages/expo-updates/ios/EXUpdates/AppLoader/AppLoaderTask.swift [hunk 31]
```diff
@@ -19,7 +18,10 @@ public protocol AppLoaderTaskDelegate: AnyObject {
    * AppLoaderTask proceed as usual.
    */
   func appLoaderTask(_: AppLoaderTask, didLoadCachedUpdate update: Update) -> Bool
-  func appLoaderTask(_: AppLoaderTask, didStartLoadingUpdate update: Update)
+  func appLoaderTask(_: AppLoaderTask, didStartCheckingForRemoteUpdate body: [String: Any])
+  func appLoaderTask(_: AppLoaderTask, didFinishCheckingForRemoteUpdate body: [String: Any])
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/AppLoader/AppLoaderTask.swift:22]: I've modified the interface to remove the unneeded parameter in `didStartCheckForRemoteUpdate()`.

I'm leaving the `didFinish` parameter as `[String: Any]` since this protocol is marked as Objective-C compatible, so I can't use a Swift struct here. If it is indeed safe to remove the Objective-C annotation, then we can do this.

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift [hunk 32]
```diff
@@ -0,0 +1,257 @@
...
+  init(changeEventDelegate: (any UpdatesStateChangeDelegate)) {
+    self.changeEventDelegate = changeEventDelegate
+  }
+
+  // MARK: - Public methods and properties
+
+  /**
+   In production, this is the AppController instance.
+   */
+  internal let changeEventDelegate: (any UpdatesStateChangeDelegate)
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:120]: Let's discuss this

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateContext.kt [hunk 33]
```diff
@@ -0,0 +1,71 @@
+package expo.modules.updates.statemachine
+
+import org.json.JSONObject
+
+/**
+The state machine context, with information intended to be consumed by application JS code.
+ */
+data class UpdatesStateContext(
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateContext.kt:8]: Done

File: packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateEvent.kt [hunk 34]
```diff
@@ -0,0 +1,54 @@
+package expo.modules.updates.statemachine
+
+import org.json.JSONObject
+
+/**
+Structure representing an event that can be sent to the machine.
+Convenience getters are provided to get derived properties that will be
+used to modify the context when the machine processes an event.
+ */
+data class UpdatesStateEvent(
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateEvent.kt:8]: Done

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift (hunk 10 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift]: Done

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift [hunk 35]
```diff
@@ -0,0 +1,306 @@
...
+ */
+internal class UpdatesStateMachine {
+  private let logger = UpdatesLogger()
+
+  // MARK: - Public methods and properties
+
+  /**
+   In production, this is the AppController instance.
+   */
+  internal var changeEventSender: (any UpdatesStateChangeEventSender)?
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift]: Done

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift [hunk 36]
```diff
@@ -0,0 +1,417 @@
...
+
+  var json: [String: Any?] {
+    return [
+      "isUpdateAvailable": self.isUpdateAvailable,
+      "isUpdatePending": self.isUpdatePending,
+      "isRollback": self.isRollback,
+      "isChecking": self.isChecking,
+      "isDownloading": self.isDownloading,
+      "isRestarting": self.isRestarting,
+      // We pass NSNulls to keep this as a [String: Any] map
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:172]: remove comment

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift (hunk 32 above)
[the remark this answers is not in the document]
Comment (wschurman) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:263]: I was under the impression that all delegates needed to be weak to prevent retain cycles (controller retains state machine retains controller). I guess weak requires it be a var and be nullable, but that's technically the most correct since we want a tear down of the controller to not be blocked by this.

**Review by wschurman — APPROVED** [review]
Just a few changes before commit.

For future PRs, I recommend splitting them into stacked PRs to improve reviewability and speed. Github is partially to blame (force-pushing clears context), but it's the tools we have to work within so I think stacked PRs generally help here. Also, I would have separated out the fix for extra params into its own PR as that took more time to review this for what was an independent fix. I appreciate you fixing it though! Nice find.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt [hunk 37]
```diff
@@ -273,6 +281,28 @@ class UpdatesController private constructor(
...
+              UpdatesStateEventType.CheckCompleteAvailable,
+              result.manifest
+            )
+          } else if (result.isRollBackToEmbedded == true) {
+            event = UpdatesStateEvent(
+              UpdatesStateEventType.CheckCompleteAvailable,
+              null,
+              null,
+              true
+            )
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesController.kt:301]: I created some convenience constructors for these events, and they are now using named parameters correctly.

File: packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.kt [hunk 38]
```diff
@@ -315,17 +352,36 @@ class UpdatesModule(
...
-        ManifestMetadata.getExtraParams(
+      try {
+        val result = ManifestMetadata.getExtraParams(
           databaseHolder.database,
           updatesServiceLocal.configuration,
         )
-      )
+        databaseHolder.releaseDatabase()
+        val resultMap = when (result) {
+          null -> Bundle()
```
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/android/src/main/java/expo/modules/updates/UpdatesModule.kt:362]: This method gets all the params, not just one, and just returns an empty map if there are no params set. The Typescript API signature expects a non-null map. So I think the empty map is the right return value.

File: packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift (hunk 32 above)
[the remark this answers is not in the document]
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:263]: At the moment, the state machine needs a delegate in order to be initialized, so the state machine gets created after the controller calls `super.init()` -- meaning the state machine property of the controller has to be nullable (weak). We can certainly do it the other way (state machine has nullable delegate, gets created by the controller before `super.init()`, then set the state machine's delegate property after).

[the remark this answers is not in the document]
Comment (wschurman) [reply packages/expo-updates/ios/EXUpdates/UpdatesStateMachine.swift:262]: Not the nullability that I'm worried about. More the retain cycles. It's still fine that it needs a delegate to be initialized, but during deinit of an object, it needs to deference all strong references and it can't do that if the object being dereferenced has a strong reference to the owning object.

https://linear.app/expo/issue/ENG-9047/weak-delegate-in-updatesstatemachine
